app/src/main/java/com/sdp13epfl2021/projStructure/AbstractProject.kt

```diff
@@ -0,0 +1,45 @@
+package com.sdp13epfl2021.projStructure
+
+
+public abstract class AbstractProject(
+        val projectName: String, val labName: String,
+        val projectOwner: String,
+        val projectDescription: String,
+        private val assignedStudents: String) {
+
+    //a companion object is the way to implement static variables in kotlin
+    companion object{
+        private final const val TAKEN: Boolean = true
+        private final const val FREE: Boolean = false
+
+    }
+    sealed class results{
+        object userCommandSuccess : results()
+        object userErrorProjectTaken : results()
+    }
+     private var isTaken = FREE;
+
+
+    fun setTaken() : results{
```

Comment:
maybe you could use [kotlin getter and setter](https://kotlinlang.org/docs/properties.html#getters-and-setters), instead of `setTaken()` and `isTaken()`

Reply:
Hi!
the kotlin data class has one main problem for my implementation : 
 the copy method we would want to use for immutability purposes does not enable us to return wrapped values and would force me to throw exeptions for error handling in the init. Exeptions are bad in kotlin because there is no way enforce exeption handling.

So I chose another design.